system/lib/wasmfs/wasmfs.cpp

```diff
@@ -0,0 +1,57 @@
+// // Copyright 2021 The Emscripten Authors.  All rights reserved.
+// // Emscripten is available under two separate licenses, the MIT license and the
+// // University of Illinois/NCSA Open Source License.  Both these licenses can be
+// // found in the LICENSE file.
+// // wasmfs.cpp will implement a new file system that replaces the existing JS filesystem.
+// // Current Status: Work in Progress.
+// // See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include <stdlib.h>
+#define __NEED_struct_iovec
+#include <emscripten/emscripten.h>
+#include <wasi/api.h>
+
+extern "C" {
+
+__wasi_errno_t __wasi_fd_write(
+  __wasi_fd_t fd, const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) {
+  // FD 1 = STDOUT and FD 2 = STDERR.
+  // Temporary hardcoding of filedescriptor values.
+  // TODO: May not want to proxy stderr (fd == 2) to the main thread.
+  // This will not not show in HTML, a console.warn in a worker is suffficient.
+  // This would be a change from the current FS.
+  if (fd == 1 || fd == 2) {
+    __wasi_size_t num = 0;
+    for (size_t i = 0; i < iovs_len; i++) {
+      const uint8_t* ptr = iovs[i].buf;
+      __wasi_size_t len = iovs[i].buf_len;
+      for (__wasi_size_t j = 0; j < len; j++) {
+        EM_ASM({ WASMFS.printChar($0, HEAPU8[$1]); }, fd, ptr + j);
+      }
+      num += len;
+    }
+    *nwritten = num;
+  }
+  return 0;
+}
+
+__wasi_errno_t __wasi_fd_seek(
+  __wasi_fd_t fd, __wasi_filedelta_t offset, __wasi_whence_t whence, __wasi_filesize_t* newoffset) {
+  EM_ASM({out($0 + ' or __wasi_fd_seek has been temporarily stubbed and is inert')},
+    __PRETTY_FUNCTION__);
+  abort();
```

Comment:
Hmm, it is unfortunate to use `EM_ASM` here actually - it increases code size to add the JS integration. But we can't rely on C++ to print because this is the actual printing code :smile: Anyhow, as we will implement these, this lgtm.

Reply:
Mmm, yes there is certainly a circular dependency 😄

Reply:
We have the lower level `emscipten_console_log` which you could use here.  (we don't have a `console_log_printf` flavor bug we could add one).

https://github.com/emscripten-core/emscripten/blob/12b2b675c202d2abc0a3f6c756939dde39eb199e/system/include/emscripten/html5.h#L500`